### PR TITLE
Add JSONObjectToYAMLObject

### PR DESCRIPTION
**Pull request by sttts**
`JSONObjectToYAMLObject` is used to convert to a YAML MapSlice representation without going through a byte slice representation, which is too heavy on allocations for some applications.

### CODE REVIEW

File: yaml_test.go [hunk 1]
```diff
@@ -421,3 +425,71 @@ foo: baz
...
+				{"map", yaml.MapSlice{{"foo", "bar"}}},
+				{"slice", []interface{}{"foo", "bar"}},
+				{"string", string("foo")},
+				{"uint64 big", uint64(1) << 63},
+			},
+		},
+	}
+	for _, tt := range tests {
+		t.Run(tt.name, func(t *testing.T) {
+			got := JSONObjectToYAMLObject(tt.input)
```
Comment (liggitt) [inline yaml_test.go]: should this also produce the same result as marshaling the json object to json bytes, then calling `yaml.Unmarshal`? if so, can we do that and compare as well?

File: yaml_test.go [hunk 2]
```diff
@@ -421,3 +425,71 @@ foo: baz
...
+				"bool":         true,
+				"float64":      float64(42.1),
+				"fractionless": float64(42),
+				"int":          int(42),
+				"int64":        int64(42),
+				"int64 big":    float64(math.Pow(2, 62)),
+				"map":          map[string]interface{}{"foo": "bar"},
+				"slice":        []interface{}{"foo", "bar"},
+				"string":       string("foo"),
+				"uint64 big":   float64(math.Pow(2, 63)),
```
[the remark this answers is not in the document]
Comment (neolit123) [reply yaml_test.go]: also what happens if we pass a big `int64(x)` value on a 32bit word size?
would the value be truncated by translating to `int()`?

File: yaml.go [hunk 3]
```diff
@@ -317,3 +317,47 @@ func convertToJSONableObject(yamlObj interface{}, jsonTarget *reflect.Value) (in
 		return yamlObj, nil
 	}
 }
+
+// JSONObjectToYAMLObject converts an in-memory JSON object into a YAML in-memory MapSlice,
+// without going through a byte representation.
+func JSONObjectToYAMLObject(j map[string]interface{}) yaml.MapSlice {
```
Comment (apelisse) [inline yaml.go]: We already have a method to go from JSonToYaml and vice-versa with a different interface since it takes bytes rather than interface. I'm OK having two interfaces, but now we also have to implementations?

File: yaml_test.go (hunk 1 above)
Comment (apelisse) [reply yaml_test.go]: Thank you Jordan, I was going to say that :-)

File: yaml.go [hunk 4]
```diff
@@ -317,3 +317,47 @@ func convertToJSONableObject(yamlObj interface{}, jsonTarget *reflect.Value) (in
 		return yamlObj, nil
 	}
 }
+
+// JSONObjectToYAMLObject converts an in-memory JSON object into a YAML in-memory MapSlice,
+// without going through a byte representation.
```
[the remark this answers is not in the document]
Comment (sttts) [reply yaml.go]: done

File: yaml.go (hunk 3 above)
Comment (sttts) [reply yaml.go]: `JSONToYAML` is doing something different, and obviously in a different way. We could implement `JSONToYAML` in terms of `json.Unmarshal` and `JSONObjectToYAMLObject`. But that would add one conversion step. I haven't checked, but I expect it to be slower than what we have.

IMO we don't have two interfaces for the same operation, but two distinct operations. `[]byte` is not in-memory YAML. In the same sense an int64 in-memory representation is not a string number.

File: yaml_test.go (hunk 1 above)
Comment (sttts) [reply yaml_test.go]: added tests and fixed some corner cases to behave the same.

File: yaml_test.go (hunk 2 above)
[the remark this answers is not in the document]
Comment (sttts) [reply yaml_test.go]: `-float64(math.Pow(2, 63)` round trips to a float64 non-precise number with `yaml.Unmarshal(json.Marshal(x))` although it fits into int64.

**Comment by sttts** [comment]
@liggitt @apelisse addressed comments.

File: yaml.go [hunk 5]
```diff
@@ -317,3 +317,63 @@ func convertToJSONableObject(yamlObj interface{}, jsonTarget *reflect.Value) (in
...
+// int64 and float64 are down casted following the logic of github.com/go-yaml/yaml:
+// - float64s are down-casted as far as possible without data-loss to int, int64, uint64.
+// - int64s are down-casted to int if possible without data-loss.
+//
+// Big int/int64/uint64 do not lose precision as in the json-yaml roundtripping case.
+//
+// string, bool and any other types are unchanged.
+func JSONObjectToYAMLObject(j map[string]interface{}) yaml.MapSlice {
+	if len(j) == 0 {
+		return nil
```
Comment (liggitt) [inline yaml.go]: this is surprising... doesn't this mean that an empty map doesn't round trip? if so, doesn't that break uses like CRD status (`status: {}`)

Comment (sttts) [reply yaml.go]: It is what the yaml library does, compare the tests.

**Comment by errordeveloper** [comment]
Is this going in any time soon?

**Comment by dims** [comment]
@liggitt can you please look over the responses from @sttts ?

File: yaml.go (hunk 5 above)
Comment (liggitt) [reply yaml.go]: hmm... it's still not quite clear to me who is intending to call this helper method, and whether this `len(0) => nil` treatment is going to bite us... any chance we can fix the yaml library instead?

Comment (sttts) [reply yaml.go]: Not that this is just the top-level conversion func. Further down we have `jsonToYAMLValue` which is faithful with empty maps.

Comment (sttts) [reply yaml.go]: @liggitt actually this semantics is not that surprising and nothing to change:

A yaml in-memory data structure has `interface{}{nil}` for null values. `yaml.MapSlice(nil)` is **not** null, but the empty map. It marshals to `{}`.

Comment (liggitt) [reply yaml.go]: Ah, ok
